worker/src/primary_connector.rs

```diff
@@ -3,94 +3,152 @@
+
+                Some(()) = futures.next() => ()
             }
         }
     }
 }
 
-#[derive(Clone)]
-pub struct PrimaryClient {
-    /// The primary network address.
-    _primary_address: Multiaddr,
-    client: WorkerToPrimaryClient<Channel>,
+pub struct WorkerToPrimaryNetwork {
+    address: Option<Multiaddr>,
+    client: Option<WorkerToPrimaryClient<Channel>>,
+    config: mysten_network::config::Config,
+    retry_config: RetryConfig,
+    executor: BoundedExecutor,
 }
 
-impl PrimaryClient {
-    pub fn new(address: Multiaddr) -> Self {
-        //TODO don't panic here if address isn't supported
-        let config = mysten_network::config::Config::new();
-        let channel = config.connect_lazy(&address).unwrap();
-        let client = WorkerToPrimaryClient::new(channel);
+impl Default for WorkerToPrimaryNetwork {
+    fn default() -> Self {
+        let retry_config = RetryConfig {
+            // Retry for forever
+            retrying_max_elapsed_time: None,
+            ..Default::default()
+        };
 
         Self {
-            _primary_address: address,
-            client,
+            address: None,
+            client: Default::default(),
+            config: Default::default(),
+            retry_config,
+            executor: BoundedExecutor::new(MAX_TASK_CONCURRENCY, Handle::current()),
         }
     }
+}
+
+impl WorkerToPrimaryNetwork {
+    pub async fn send<PublicKey: VerifyingKey>(
+        &mut self,
+        address: Multiaddr,
+        message: &WorkerPrimaryMessage<PublicKey>,
+    ) -> CancelHandler<()> {
+        let new_client = match &self.address {
+            None => true,
+            Some(x) if x != &address => true,
+            _ => false,
+        };
+        if new_client {
+            let channel = self.config.connect_lazy(&address).unwrap();
+            self.client = Some(WorkerToPrimaryClient::new(channel));
+            self.address = Some(address);
+        }
 
-    pub async fn send(&mut self, message: &WorkerPrimaryMessage) -> anyhow::Result<()> {
-        let message = BincodeEncodedPayload::try_from(message)?;
-        self.client
-            .send_message(message)
-            .await
-            .map_err(Into::into)
-            .map(|_| ())
+        let message =
+            BincodeEncodedPayload::try_from(message).expect("Failed to serialize payload");
+        let client = self.client.as_mut().unwrap().clone();
+        let handle = self
+            .executor
+            .spawn(
+                self.retry_config
+                    .retry(move || {
+                        let mut client = client.clone();
+                        let message = message.clone();
+                        async move { client.send_message(message).await.map_err(Into::into) }
+                    })
+                    .map(|response| {
+                        response.expect("we retry forever so this shouldn't fail");
+                    }),
+            )
+            .await;
```

Comment:
This is effectively bypassing the changes brought in in #559 and should rather build on top of them. We've fixed some nasty behavior there, so that PR may need review.

Reply:
Ah! I though this PR would be merged before #559

Reply:
let me see how to fix that

Reply:
fixed